File: Analyse exploratoire/TD1/Python/TD1.py
import pandas as pd
import numpy as np

from scipy.stats import chi2_contingency

# Importing the csv and loading it in a dataframe.
CSV_FILE_PATH = "~/esiea/Analyse exploratoire/TD1/data/menu.csv"
MENU_DF = pd.read_csv(CSV_FILE_PATH)

# Similar to R's 'summary' function
print(MENU_DF.describe())


def chisq_of_df_cols(input_df, col_1, col_2):
    """
    Computes the chi square value for a dataframe based on two column names.
    Returns this value.
    """
    groupsizes = input_df.groupby([col_1, col_2]).size()
    ctsum = groupsizes.unstack(col_1)

    return chi2_contingency(ctsum.fillna(0))


def chisq_dataframe_of_dataframe(input_df, cols):
    n = len(cols)

    chi2_results_df = pd.DataFrame(len(cols) * [len(cols) * [0]], columns=cols)

    for i in range(n - 1):
        for j in range(i + 1, n):
            chi2 = chisq_of_df_cols(
                MENU_DF, cols[i], cols[j])[0]

            chi2_results_df.iloc[i, j] = chi2
            chi2_results_df.iloc[j, i] = chi2

    return chi2_results_df


def scale_column_df(input_df, norm_df, col_name):
    """
    Returns a dataframe where the specified col_name is the scaled version of the input dataframe's.
    """
    try:
        norm_df[col_name] = (
            input_df[col_name] - input_df[col_name].mean()) / input_df[col_name].std()
        return norm_df, None
    except Exception as exc:
        return norm_df, exc

def scale_df(input_df, col_names):
    norm_df = pd.DataFrame()

    for col_name in col_names:
        norm_df, exc = scale_column_df(input_df, norm_df, col_name)

        if exc is not None:
            print(exc)
    
    return norm_df

def main():
    fields_of_interest = ["Calories", "Total Fat",
                          "Cholesterol", "Sodium", "Sugars", "Protein"]

    chi2_results_df = chisq_dataframe_of_dataframe(MENU_DF, fields_of_interest)

    chi2_results_df["index"] = fields_of_interest
    chi2_results_df = chi2_results_df.set_index("index")

    print(chi2_results_df)

    chi2_results_df = chisq_dataframe_of_dataframe(scale_df(MENU_DF, fields_of_interest), fields_of_interest)

    chi2_results_df["index"] = fields_of_interest
    chi2_results_df = chi2_results_df.set_index("index")
    
    print(chi2_results_df)

if __name__ == '__main__':
    main()

File: Analyse exploratoire/TD1/TD1.R
mcdo_df <- read.csv("~/esiea/Analyse exploratoire/TD1/data/menu.csv")
summary(mcdo_df)

cor(mcdo_df$Calories, mcdo_df$Total.Fat) # On calcule la corrélation entre les deux variables.

cor_of_df <- function(df, fields_of_interest) {
  "
  Function that computes the correlation for each couple of columns in a dataframe
  and returns a matrix containing the results.
  
  The computation takes into account the fact that the matrix is symmetrical, and
  that a correlation on the same vectors is 1.
  "
  n <- length(fields_of_interest)
  
  # Initializing the matrix containing the chi square values.
  cor_matrix <- matrix(1,
                       ncol = n,
                       nrow = n,
                       dimnames = list(fields_of_interest, fields_of_interest))
  
  # Computing the Chi square values, and using the symmetry of the matrix.
  for (i in 1:(n - 1)) {
    for (j in (i + 1):n) {
      temp_cor <- cor(df[, fields_of_interest[i]],
                      df[, fields_of_interest[j]])
      
      cor_matrix[i, j] <- temp_cor
      cor_matrix[j, i] <- temp_cor
    }
  }
  
  return(cor_matrix)
}


# Name of the variables we are interested in computing the Chi square test.
fields_of_interest <- c("Calories", "Total.Fat",
                        "Cholesterol", "Sodium", "Sugars", "Protein")

# Compute the chi square matrix.
(correlation_data_frame <- cor_of_df(mcdo_df[fields_of_interest], fields_of_interest))


# Same, but after scaling
(correlation_scaled_data_frame <- cor_of_df(scale(mcdo_df[fields_of_interest]), fields_of_interest))


# Difference between the two
correlation_data_frame - correlation_scaled_data_frame

correlation_data_frame>0.5

shapiro_of_df <- function(df, fields_of_interest) {
  "
  Function that computes the shapiro test's p value for each couple of columns in a dataframe
  and returns a matrix containing the results.
  
  The computation takes into account the fact that the matrix is symmetrical, and
  that a correlation on the same vectors is 1.
  "
  n <- length(fields_of_interest)
  
  # Initializing the matrix containing the chi square values.
  shapiro_array <- rep(NA, n)
  
  # Computing the Chi square values, and using the symmetry of the matrix.
  for (i in 1:n) {
    shapiro_array[i] <- shapiro.test(df[, fields_of_interest[i]])$p.value
  }
  
  return(shapiro_array)
}

(shapiro_of_df(mcdo_df, fields_of_interest)<0.05)

library(rgl)

list <- c("Calories", "Total.Fat", "Cholesterol")

plot3d(mcdo_df$Calories,
       mcdo_df$Total.Fat,
       mcdo_df$Cholesterol,
       type="s")

interesting_df <- scale(mcdo_df[, c("Calories", "Total.Fat", "Cholesterol")])

lims <- c(min(interesting_df),
          max(interesting_df))

plot3d(interesting_df,
       type = "s")

plot3d(ellipse3d(cor(cbind(interesting_df$Calories,
                           interesting_df$Total.Fat,
                           interesting_df$Cholesterol))),
                           col="grey",
                          add=TRUE)

library(ade4)
list <- c("Calories", "Total.Fat", "Cholesterol")
(acp <- dudi.pca(scale(mcdo_df[, list]),
                 center=FALSE,
                 scale=FALSE,
                 scannf = FALSE,
                 nf = 3))

names(acp)


pve <- 100 * acp$eig / sum(acp$eig) # On calcule le pourcentage d'intertie de chaque vecteur propre.
cumsum(pve)

(inertie <-inertia.dudi(acp, col.inertia=TRUE))

round(acp$co ,2)
s.corcircle(acp$co,
            xax=1,
            yax=2)


s.label(acp$li,
        xax = 1,
        yax = 2,
        label=as.character(mcdo_df$Item),
        clabel=1.5)

gcol <- c("red1", "red4","orange")
s.class(dfxy = acp$li, fac = mcdo_df$Category, col = gcol, xax =
          1, yax = 2)

scatter(acp)

list <- c("Calories", "Total.Fat", "Cholesterol", "Sodium", "Sugars", "Protein")
(acp <- dudi.pca(scale(mcdo_df[, list]),
                 center=FALSE,
                 scale=FALSE,
                 scannf = FALSE,
                 nf = 3))
(inertie <-inertia.dudi(acp, col.inertia=TRUE))

round(acp$co ,2)
s.corcircle(acp$co,
            xax=1,
            yax=2)
